vicky/src/lib/database/entities/task.rs

```diff
@@ -0,0 +1,415 @@
+                TaskStatus::RUNNING => "RUNNING",
+                TaskStatus::FINISHED(r) => match r {
+                    TaskResult::SUCCESS => "FINISHED::SUCCESS",
+                    TaskResult::ERROR => "FINISHED::ERROR",
+                },
+            };
+            write!(f, "{}", str)
+        }
+    }
+
+    impl From<&str> for TaskStatus {
+        fn from(str: &str) -> TaskStatus {
+            match str {
+                "RUNNING" => TaskStatus::RUNNING,
+                "FINISHED::SUCCESS" => TaskStatus::FINISHED(TaskResult::SUCCESS),
+                "FINISHED::ERROR" => TaskStatus::FINISHED(TaskResult::ERROR),
+                _ => TaskStatus::NEW,
+            }
+        }
+    }
+
+    impl From<&Task> for DbTask {
+        fn from(task: &Task) -> DbTask {
+            DbTask {
+                id: task.id,
+                display_name: task.display_name.clone(),
+                status: task.status.to_string(),
+                features: task.features.join("||"),
+                flake_ref_uri: task.flake_ref.flake.clone(),
+                flake_ref_args: task.flake_ref.args.join("||"),
+            }
+        }
+    }
+
+    #[derive(Selectable, Identifiable, Queryable, Debug)]
+    #[diesel(table_name = locks)]
+    struct DbLock {
+        id: i32,
+        task_id: Uuid,
+        name: String,
+        type_: String,
+    }
+
+    #[derive(Insertable, Debug)]
+    #[diesel(table_name = locks)]
+    struct NewDbLock {
+        task_id: Uuid,
+        name: String,
+        type_: String,
+    }
+
+    impl From<DbLock> for NewDbLock {
+        fn from(value: DbLock) -> Self {
+            NewDbLock {
+                task_id: value.task_id,
+                name: value.name,
+                type_: value.type_,
+            }
+        }
+    }
+
+    impl DbLock {
+        fn from_lock(lock: &Lock, task_id: Uuid) -> Self {
+            match lock {
+                Lock::WRITE { name } => DbLock {
+                    id: -1,
+                    task_id,
+                    name: name.clone(),
+                    type_: "WRITE".to_string(),
+                },
+                Lock::READ { name } => DbLock {
+                    id: -1,
+                    task_id,
+                    name: name.clone(),
+                    type_: "READ".to_string(),
+                },
+            }
+        }
+    }
+
+    impl From<DbLock> for Lock {
+        fn from(lock: DbLock) -> Lock {
+            match lock.type_.as_str() {
+                "WRITE" => Lock::WRITE { name: lock.name },
+                "READ" => Lock::READ { name: lock.name },
+                _ => panic!(
+                    "Can't parse lock from database lock. Database corrupted? \
+                Expected READ or WRITE but found {} as type at key {}.",
+                    lock.type_, lock.id
+                ),
+            }
+        }
+    }
+
+    #[database("postgres_db")]
+    pub struct Database(diesel::PgConnection);
```

Comment:
Is there a reason, why this is in here? 
If this is a more general thing, then we should move it out.
If it is needed for whatever reason, a comment would be nice!

Reply:
It's a wrapper that encapsulates the database connection for connection pooling. It's a rocket thing.

`#[database("postgres_db")]` is a macro that gives the Database struct connection pooling abilities. It takes "postgres_db" from the Rocket.toml and uses that connection string to initialize its connections. The rocket documentation explains this really well too and a quick look at the macro also shows this.

Reply:
Might move it somewhere else tho. I am also not too keen on the structure (which i already improved since before). Will check again.